File: sentinel/src/main/scala/pl/edu/agh/sentinel/events/TaskEvent.scala
package pl.edu.agh.sentinel.events

import zio.json
import zio.json._
import zio.json.{ DeriveJsonCodec, DeriveJsonDecoder, DeriveJsonEncoder, JsonCodec, JsonDecoder, JsonEncoder }

import java.time.Instant

@jsonDiscriminator("type")
sealed trait TaskEvent {
  def taskId: String;

  def teamId: String;

  def timestamp: Instant
}

object TaskEvent {
  final case class TaskCreated(
    taskId: String,
    title: String,
    teamId: String,
    columnId: String,
    creatorId: String,
    timestamp: Instant,
  ) extends TaskEvent derives JsonCodec

  final case class TaskAssigned(
    taskId: String,
    assigneeId: String,
    teamId: String,
    projectId: String,
    timestamp: Instant,
  ) extends TaskEvent derives JsonCodec

  final case class TaskClosed(
    taskId: String,
    title: String,
    teamId: String,
    columnId: String,
    creatorId: String,
    timestamp: Instant,
  ) extends TaskEvent derives JsonCodec

  final case class TaskMoved(
    taskId: String,
    teamId: String,
    fromColumnId: String,
    toColumnId: String,
    movedBy: String,
    timestamp: Instant,
  ) extends TaskEvent derives JsonCodec

  given JsonCodec[TaskEvent] = DeriveJsonCodec.gen[TaskEvent]
}

File: io-project/src/main/java/pl/edu/agh/io_project/kafka/TaskEvent/TaskEvent.java
package pl.edu.agh.io_project.kafka.TaskEvent;


import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.annotation.JsonTypeName;

import java.time.Instant;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
@JsonSubTypes({
        @JsonSubTypes.Type(value = TaskCreated.class, name = "TaskCreated"),
        @JsonSubTypes.Type(value = TaskAssigned.class, name = "TaskAssigned"),
        @JsonSubTypes.Type(value = TaskClosed.class, name = "TaskClosed"),
        @JsonSubTypes.Type(value = TaskMoved.class, name = "TaskMoved"),
})
public sealed interface TaskEvent permits TaskCreated, TaskAssigned, TaskClosed, TaskMoved {
    String getTaskId();
    Instant getTimestamp();
}


